lib/goal/cache/goalCaching.ts

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright © 2019 Atomist, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+    GitProject,
+    Project,
+    RepoRef,
+} from "@atomist/automation-client";
+import { gatherFromFiles } from "@atomist/automation-client/lib/project/util/projectUtils";
+import {
+    AnyPush,
+    ExecuteGoalResult,
+    GoalInvocation,
+    GoalProjectListener,
+    GoalProjectListenerEvent,
+    GoalProjectListenerRegistration,
+    ProgressLog,
+    PushTest,
+} from "@atomist/sdm";
+import * as _ from "lodash";
+
+export interface GoalArtifactCache {
+    putInCache(id: RepoRef, project: Project, file: string[], log: ProgressLog): Promise<void>;
+    retrieveFromCache(id: RepoRef, project: Project, log: ProgressLog): Promise<void>;
+    removeFromCache(id: RepoRef): Promise<void>;
+}
+
+const defaultGoalArtifactCacheOptions: GoalArtifactCacheOptions = {
+    pushTest: AnyPush,
+    globPattern: "**/*.jar",
+    fallbackListenerOnCacheMiss: () => { throw Error("No entry found in cache"); },
+};
+
+export interface GoalArtifactCacheOptions {
+    pushTest: PushTest;
+    globPattern: string;
+    fallbackListenerOnCacheMiss: GoalProjectListener;
+}
+
+export function cacheGoalArtifacts(artifactArchiveCacher: GoalArtifactCache,
+                                   options: Partial<GoalArtifactCacheOptions> = {}): GoalProjectListenerRegistration {
```

Comment:
We don't use Partials like this on public APIs. The `GoalArtifactCacheOptions` should define optional properties directly in the interface.